src/shogun/clustering/KMeansBase.cpp

```diff
@@ -130,6 +130,44 @@ void KMeansBase::compute_cluster_variances()
 	}
 }
 
+SGMatrix<float64_t> KMeansBase::compute_std_dev() const
+{
+	require(cluster_centers.size() > 0, "KMeans is not trained!");
+
+	SGMatrix<float64_t> points = distance->get_rhs()
+	                                 ->as<DenseFeatures<float64_t>>()
+	                                 ->get_feature_matrix();
+	SGVector<float64_t> cluster_assignments = const_cast<KMeansBase*>(this)
+	                                              ->apply()
+	                                              ->as<MulticlassLabels>()
+	                                              ->get_labels();
+
+	SGVector<int32_t> counts(k);
+	SGMatrix<float64_t> means = cluster_centers.clone();
+	SGMatrix<float64_t> squares_sums(dimensions, k);
+
+	for (int32_t point_number : range(cluster_assignments.vlen))
+	{
+		auto cluster_number = (int32_t) cluster_assignments[point_number];
+		auto point = points.get_column(point_number);
+		auto& count = counts[cluster_number];
+		auto mean = means.get_column(cluster_number);
+		auto squares_sum = squares_sums.get_column(cluster_number);
+
+		count += 1;
+		auto delta1 = linalg::add(point, mean, 1., -1.);
+		linalg::add(mean, linalg::scale(delta1, 1. / count), mean);
```

Comment:
Can we compute the mean differently @karlnapf? The issue is just that you are doing a division in a loop.. Is there no geometric series that could be cheaper to compute?

Reply:
this is from https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm

@vinnik-dmitry07 instead of implementing this online algorithm in here, could you move this somewhere else, so that it is also usable from other parts of the code? I envision an updater being instantiated and then repeatedly being called, a bit like an iterator
 @gf712 any ideas where would be best and what the API would look like?

Reply:
Sure, we would just have a class with a update(datapoint) public class function. Then can have some derived classes for mean and variance (unless you want to combine them in a single class). Not sure what other algos would be useful to have online update for?

Reply:
yes definitely useful! The class should support both mean and variance but each should be optional (one wants one without the other one)